### PR TITLE
[INTEGRATION[spark] add visitors for Hive

### PR DESCRIPTION
**Pull request by mobuchowski**
Implement visitors for `InsertIntoHiveTable` and `InsertIntoHiveDirCommand`.

Solves https://github.com/OpenLineage/OpenLineage/issues/187.

Signed-off-by: Maciej Obuchowski <maciej.obuchowski@getindata.com>

### CODE REVIEW

**Review by OleksandrDvornik — APPROVED** [review]

File: client/java/generator/.classpath [hunk 1]
```diff
@@ -26,7 +26,6 @@
 			<attribute name="test" value="true"/>
 		</attributes>
 	</classpathentry>
-	<classpathentry kind="con" path="org.eclipse.jdt.launching.JRE_CONTAINER/org.eclipse.jdt.internal.debug.ui.launcher.StandardVMType/JavaSE-13/"/>
-	<classpathentry kind="con" path="org.eclipse.buildship.core.gradleclasspathcontainer"/>
+	<classpathentry kind="con" path="org.eclipse.jdt.launching.JRE_CONTAINER/org.eclipse.jdt.internal.debug.ui.launcher.StandardVMType/JavaSE-13/"/>	<classpathentry kind="con" path="org.eclipse.buildship.core.gradleclasspathcontainer"/>
```
Comment (OleksandrDvornik) [inline client/java/generator/.classpath:29]: why do we need this file?

**Review by OleksandrDvornik — COMMENTED** [review]
we have a PR with refactoring https://github.com/OpenLineage/OpenLineage/pull/263
I guess it should go first

File: client/java/generator/.classpath (hunk 1 above)
Comment (mobuchowski) [reply client/java/generator/.classpath]: Not sure, probably we should remove it. 
Anyway, change to this file was commited accidentally, removed it.

File: integration/spark/src/main/java/io/openlineage/spark/agent/lifecycle/plan/InsertIntoHiveDirVisitor.java [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+import io.openlineage.spark.agent.util.PlanUtils;
+import java.net.URI;
+import java.util.Collections;
+import java.util.List;
+import org.apache.spark.sql.catalyst.catalog.CatalogStorageFormat;
+import org.apache.spark.sql.catalyst.plans.logical.LogicalPlan;
+import org.apache.spark.sql.hive.execution.InsertIntoHiveDirCommand;
+import scala.runtime.AbstractFunction0;
+
+public class InsertIntoHiveDirVisitor extends QueryPlanVisitor<InsertIntoHiveDirCommand> {
```
[the remark this answers is not in the document]
Comment (mobuchowski) [reply integration/spark/src/main/java/io/openlineage/spark/agent/lifecycle/plan/InsertIntoHiveDirVisitor.java:17]: Done.

File: integration/spark/src/main/java/io/openlineage/spark/agent/lifecycle/plan/InsertIntoHiveTableVisitor.java [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+ * {@link LogicalPlan} visitor that matches an {@link InsertIntoHiveTable} and extracts the output
+ * {@link OpenLineage.Dataset} being written.
+ */
+public class InsertIntoHiveTableVisitor extends QueryPlanVisitor<InsertIntoHiveTable> {
+
+  @Override
+  public List<OpenLineage.Dataset> apply(LogicalPlan x) {
+    InsertIntoHiveTable cmd = (InsertIntoHiveTable) x;
+    CatalogTable table = cmd.table();
+    return Collections.singletonList(PlanUtils.getDataset(table.location(), table.schema()));
```
[the remark this answers is not in the document]
Comment (mobuchowski) [reply integration/spark/src/main/java/io/openlineage/spark/agent/lifecycle/plan/InsertIntoHiveTableVisitor.java]: Fixed.

File: integration/spark/src/test/java/io/openlineage/spark/agent/SparkContainerIntegrationTest.java [hunk 4]
```diff
@@ -207,4 +205,33 @@ public void testPysparkRddToTable() throws IOException, InterruptedException {
             .withPath("/api/v1/lineage")
             .withBody(json(completeTableEvent, MatchType.ONLY_MATCHING_FIELDS)));
   }
+
+  @Test
+  public void testPysparkSQLHiveTest() throws IOException, InterruptedException {
```
[the remark this answers is not in the document]
Comment (mobuchowski) [reply integration/spark/src/test/java/io/openlineage/spark/agent/SparkContainerIntegrationTest.java]: Also, added test for `INSERT OVERWRITE`.

File: integration/spark/src/test/java/io/openlineage/spark/agent/SparkContainerIntegrationTest.java [hunk 5]
```diff
@@ -94,6 +94,24 @@ private static MockServerContainer makeMockServerContainer() {
         .withCommand(command);
   }
 
+  private static GenericContainer<?> makePysparkContainerWithDefaultConf(
+      String namespace, String command) {
+    return makePysparkContainer(
+        "--master",
+        "local",
+        "--conf",
+        "spark.openlineage.host=" + "http://openlineageclient:1080",
```
Comment (wslulciuc) [inline integration/spark/src/test/java/io/openlineage/spark/agent/SparkContainerIntegrationTest.java:103]: _Minor_: Mind if we use a more reasonable http port like `8080`? (used as the default API port for Marquez)

File: integration/spark/src/test/java/io/openlineage/spark/agent/SparkContainerIntegrationTest.java [hunk 6]
```diff
@@ -112,99 +130,142 @@ private static void consumeOutput(org.testcontainers.containers.output.OutputFra
...
-            "spark.openlineage.url="
-                + "http://openlineageclient:1080/api/v1/namespaces/testPysparkWordCountWithCliArgs",
-            "--conf",
-            "spark.extraListeners=" + OpenLineageSparkListener.class.getName(),
-            "--jars",
-            "/opt/libs/" + System.getProperty("openlineage.spark.jar"),
-            "/opt/spark_scripts/spark_word_count.py");
-    pyspark.setWaitStrategy(Wait.forLogMessage(".*ShutdownHookManager: Shutdown hook called.*", 1));
-    pyspark.start();
+  // @Test
```
Comment (wslulciuc) [inline integration/spark/src/test/java/io/openlineage/spark/agent/SparkContainerIntegrationTest.java:133]: Do we want to comment out these tests?

File: integration/spark/src/test/java/io/openlineage/spark/agent/SparkContainerIntegrationTest.java (hunk 5 above)
Comment (mobuchowski) [reply integration/spark/src/test/java/io/openlineage/spark/agent/SparkContainerIntegrationTest.java:103]: This is connecting to MockServer, and I unfortunately don't see easy way to change that via env variables. 

https://www.mock-server.com/mock_server/configuration_properties.html

File: integration/spark/src/test/java/io/openlineage/spark/agent/SparkContainerIntegrationTest.java (hunk 6 above)
Comment (mobuchowski) [reply integration/spark/src/test/java/io/openlineage/spark/agent/SparkContainerIntegrationTest.java]: Fixed.

File: integration/spark/src/test/java/io/openlineage/spark/agent/SparkContainerIntegrationTest.java (hunk 5 above)
Comment (wslulciuc) [reply integration/spark/src/test/java/io/openlineage/spark/agent/SparkContainerIntegrationTest.java:103]: Ahh, actually, that makes total sense. You can disregard this comment ;)

File: integration/spark/src/main/java/io/openlineage/spark/agent/lifecycle/plan/InsertIntoHiveTableVisitor.java [hunk 7]
```diff
@@ -0,0 +1,40 @@
...
+      path = new Path(table.location());
+      if (table.location().getScheme() == null) {
+        path = new Path("file", null, table.location().toString());
+      }
+    } catch (Exception e) { // Java does not recognize scala exception
+      if (e instanceof AnalysisException) {
+        path = new Path(table.qualifiedName());
+        if (table.location().getScheme() == null) {
+          path = new Path("file", null, table.location().toString());
+        }
```
Comment (collado-mike) [inline integration/spark/src/main/java/io/openlineage/spark/agent/lifecycle/plan/InsertIntoHiveTableVisitor.java:34]: I'd likely default to a Dataset with something like `hive://<metastore.uri>/<qualified_name>`. You should be able to get the `metastore.uri` from the spark configuration `spark.sql.hive.metastore.uris`

**Review by collado-mike — APPROVED** [review]

File: integration/spark/src/main/java/io/openlineage/spark/agent/lifecycle/plan/InsertIntoHiveTableVisitor.java (hunk 7 above)
Comment (mobuchowski) [reply integration/spark/src/main/java/io/openlineage/spark/agent/lifecycle/plan/InsertIntoHiveTableVisitor.java]: Done.

**Pull request merged**